File: Algorithms/beautiful-arrangement.cpp
// 526. Beautiful Arrangement
// https://leetcode.com/problems/beautiful-arrangement/

/*
Suppose you have N integers from 1 to N. We define a beautiful arrangement as an array that is constructed by these N numbers successfully if one of the following is true for the ith position (1 ? i ? N) in this array:

The number at the ith position is divisible by i.
i is divisible by the number at the ith position.
Now given N, how many beautiful arrangements can you construct?

Example 1:
Input: 2
Output: 2
Explanation: 

The first beautiful arrangement is [1, 2]:

Number at the 1st position (i=1) is 1, and 1 is divisible by i (i=1).

Number at the 2nd position (i=2) is 2, and 2 is divisible by i (i=2).

The second beautiful arrangement is [2, 1]:

Number at the 1st position (i=1) is 2, and 2 is divisible by i (i=1).

Number at the 2nd position (i=2) is 1, and i (i=2) is divisible by 1.
Note:
N is a positive integer and will not exceed 15.
*/

class Solution {
public:
	int countArrangement(int N) {
		vector<int> A(N);
		iota(begin(A), end(A), 1);
		return f(A, N, 0);
	}
private:
	int f(vector<int> & A, int N, int start) {
		if (start >= N) {
			return 1;
		}
		int result = 0;
		for (int i = start; i < N; ++i) {
			swap(A[start], A[i]);
			if ((start + 1) % A[start] == 0 or A[start] % (start + 1) == 0) {
				result += f(A, N, start + 1);
			}
			swap(A[start], A[i]);
		}
		return result;
	}
};

class Solution {
public:
	int countArrangement(int N) {
		vector<bool> A(N, true);
		return f(A, N, 0);
	}
private:
	int f(vector<bool> & A, int N, int start) {
		if (start >= N) {
			return 1;
		}
		int result = 0;
		for (int i = 0; i < N; ++i) {
			if (A[i] and ((start + 1) % (i + 1) == 0 or (i + 1) % (start + 1) == 0)) {
				A[i] = false;
				result += f(A, N, start + 1);
				A[i] = true;
			}
		}
		return result;
	}
};

File: Algorithms/beautiful-arrangement.py
526. Beautiful Arrangement
https://leetcode.com/problems/beautiful-arrangement/

Suppose you have n integers labeled 1 through n. A permutation of those n integers perm (1-indexed) is considered a beautiful arrangement if for every i (1 <= i <= n), either of the following is true:

perm[i] is divisible by i.
i is divisible by perm[i].
Given an integer n, return the number of the beautiful arrangements that you can construct.

 

Example 1:

Input: n = 2
Output: 2
Explanation: 
The first beautiful arrangement is [1,2]:
    - perm[1] = 1 is divisible by i = 1
    - perm[2] = 2 is divisible by i = 2
The second beautiful arrangement is [2,1]:
    - perm[1] = 2 is divisible by i = 1
    - i = 2 is divisible by perm[2] = 1
Example 2:

Input: n = 1
Output: 1
 

Constraints:

1 <= n <= 15

class Solution:
    def countArrangement(self, n: int) -> int:
        A = [set() for i in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if i % j == 0 or j % i == 0:
                    A[i].add(j)
        B = [True for i in range(n + 1)]
        
        def f1(start):
            if start > n:
                return 1
            result = 0
            for i in A[start]:
                if B[i]:
                    B[i] = False
                    result += f1(start + 1)
                    B[i] = True
            return result
        
        return f1(1)

if __name__ == "__main__":
    sol = Solution()
    n = 15
    A = [0 for i in range(n)]
    for i in range(n):
        A[i] = sol.countArrangement(i + 1)
    print(A)